File: old/2020/Day15.hs
{-# LANGUAGE TemplateHaskell #-}
import Control.Lens
import Control.Monad.ST
import Control.Monad.State.Strict
import Data.HashTable.ST.Basic as M

-- given an int, return the
type History = Either Int (Int, Int)
data Game s = Game
  { _turn :: Int
  , _history :: HashTable s Int History
  , _lastSpoken :: Int
  } deriving Show
makeLenses ''Game

updateHistory :: Int -> StateT (Game s) (ST s) ()
updateHistory n game = do
  index <- gets _turn
  history %= (M.mutate (((, ()) . Just . f index) n)
  where
    f :: Int -> Maybe History -> History
    f index = \case
      Nothing                     -> Left index
      Just (Left prevIndex)       -> Right (index, prevIndex)
      Just (Right (prevIndex, _)) -> Right (index, prevIndex)

startWith :: Int -> StateT (Game s) (ST s) ()
startWith n game = do
  updateHistory n
  lastSpoken .= n
  turn += 1

nextTurn :: StateT (Game s) (ST s) ()
nextTurn = do
  last <- use lastSpoken
  h <- use (history . to (! last)) -- (history . ix last))
  let next = case h of {
    Left _ -> 0;
    Right (p, q) -> p - q
  }
  updateHistory next
  lastSpoken .= next
  turn += 1

runGame :: [Int] -> Int -> ST s Game
runGame initial numTurns = do
  history <- new
  execStateT (do
    traverse startWith initial
    replicateM_ (numTurns - (length initial)) nextTurn)
    history

main :: IO ()
main = do
  let puzzleInput = [8,13,1,0,18,9]
  let example = [0, 3, 6]
  let prob1 = runGame puzzleInput 2020
  let prob2 = runGame puzzleInput 300000
  print . _lastSpoken $ prob1
  print . _lastSpoken $ runGame puzzleInput 300000
  print . _lastSpoken $ runGame puzzleInput 400000
  print . _lastSpoken $ runGame puzzleInput 500000
  print . _lastSpoken $ runGame puzzleInput 600000
  print . _lastSpoken $ runGame puzzleInput 700000
  print . _lastSpoken $ runGame puzzleInput 800000
  print . _lastSpoken $ runGame puzzleInput 900000

File: old/2020/Day15.java
class Day15 {
    static int turn = 0;
    static int last_spoken = 0;
    //static puzzle_input = [8,13,1,0,18,9]
    static int puzzle_input[] = {0, 3, 6};
    static int end_turn = 10;
    static int history[][] = new int[end_turn][2];

    public static void update_history(int n) {
        int a[] = history[n];
        if (a[0] == -1)
            history[n] = new int[]{turn, -1};
        else
            history[n] = new int[]{turn, a[0]};
    }

    public static void main(String args[]) {
        for (int i = 0; i < end_turn; i++)
            history[i] = new int[]{-1, -1};

        for (int n : puzzle_input) {
            update_history(n);
            last_spoken = n;
            System.out.println(last_spoken);
            turn++;
        }

        while (turn < end_turn) {
            int h[] = history[last_spoken];
            int next_spoken = (h[0] != -1 ? h[0] - h[1] : 0);
            last_spoken = next_spoken;
            update_history(last_spoken);
            System.out.println(last_spoken);
            turn++;
        }

        //System.out.println(last_spoken);
    }
}
